compiler/include/ForallStmt.h

```diff
@@ -70,20 +78,24 @@ class ForallStmt : public Stmt
 
   bool hasVectorizationHazard() const;
   void setHasVectorizationHazard(bool v);
+
 private:
   bool           fZippered;
   AList          fIterVars;
   AList          fIterExprs;
   AList          fShadowVars;  // may be empty
   BlockStmt*     fLoopBody;    // always present
-  bool           fFromForLoop; // see comment below
   bool           fVectorizationHazard;
 
   ForallStmt(bool zippered, BlockStmt* body);
 
 public:
-  LabelSymbol*   fContinueLabel;     // update_symbols() needs the labels
+  LabelSymbol*   fContinueLabel;
   LabelSymbol*   fErrorHandlerLabel;
+  bool           fFromForLoop; // see comment below
+  bool           fFromReduce;
+  bool           fAllowSerialIterator;
+  bool           fRequireSerialIterator;
```

Comment:
Hopefully your creation of ForallStmts is factored a little bit into functions so that creating one from a ReduceExpr (or another source) sets e.g. fAllowSerialIterator appropriately. In that case maybe you only need functions returning these, and the fields themselves could be private?